src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.PendingBranchesCollection.cs

```diff
@@ -0,0 +1,134 @@
+        {
+            private ArrayBuilder<PendingBranch> _branches;
+            private PooledDictionary<LabelSymbol, ArrayBuilder<PendingBranch>>? _branchesToLabel;
+
+            internal PendingBranchesCollection()
+            {
+                _branches = ArrayBuilder<PendingBranch>.GetInstance();
+            }
+
+            internal void Free()
+            {
+                _branches.Free();
+                FreeBranchesToLabel();
+            }
+
+            internal void Clear()
+            {
+                _branches.Clear();
+                FreeBranchesToLabel();
+            }
+
+            private void FreeBranchesToLabel()
+            {
+                if (_branchesToLabel is { })
+                {
+                    foreach (var branches in _branchesToLabel.Values)
+                    {
+                        branches.Free();
+                    }
+                    _branchesToLabel.Free();
+                    _branchesToLabel = null;
+                }
+            }
+
+            internal ImmutableArray<PendingBranch> ToImmutable()
+            {
+                return _branchesToLabel is null ?
+                    _branches.ToImmutable() :
+                    ImmutableArray.CreateRange(this);
+            }
+
+            internal ArrayBuilder<PendingBranch> GetAndRemoveBranches(LabelSymbol? label)
+            {
+                ArrayBuilder<PendingBranch>? result;
+                if (label is null)
+                {
+                    result = _branches;
+                    _branches = ArrayBuilder<PendingBranch>.GetInstance();
+                }
+                else if (_branchesToLabel is { } && _branchesToLabel.TryGetValue(label, out result))
+                {
+                    _branchesToLabel.Remove(label);
+                }
+                else
+                {
+                    result = ArrayBuilder<PendingBranch>.GetInstance();
+                }
+                return result;
+            }
+
+            internal void Add(PendingBranch branch)
+            {
+                var label = branch.Label;
+                if (label is null)
+                {
+                    _branches.Add(branch);
+                }
+                else
+                {
+                    if (_branchesToLabel is null)
+                    {
+                        _branchesToLabel = PooledDictionary<LabelSymbol, ArrayBuilder<PendingBranch>>.GetInstance();
+                    }
+                    if (!_branchesToLabel.TryGetValue(label, out var branches))
+                    {
+                        branches = ArrayBuilder<PendingBranch>.GetInstance();
+                        _branchesToLabel.Add(label, branches);
+                    }
+                    branches.Add(branch);
+                }
+            }
+
+            internal void AddRange(PendingBranchesCollection collection)
+            {
```

Comment:
Consider adding the following as it could avoid intermediate array allocations


```suggestion
            {
                _branches.EnsureCapcity(_branches.Count + collection._branches.Count); 
```

Reply:
Updated to use `ArrayBuilder<T>.AddRange()` directly.